Vault unit does not re-join raft cluster after pod crashing

#### Describe the bug

There seems to be a bug where when a pod gets deleted, it does not properly join back the raft cluster. The node that crashed does not show up in the peer list with its new IP address.

In the example below, the initial IP address of the deleted node was `10.1.182.57` and the new one it got was `10.1.182.19`.

#### To Reproduce

1. `juju deploy vault-k8s -n 3 --trust --channel=edge`
2. `kubectl delete pod vault-k8s-1 -n <my model>
3. `vault operator raft list-peers`

#### Expected behavior

The node that crashed joins the raft cluster and the correct IP address is shown when running `vault operator raft list-peers`.

#### Logs


After the pod crashed and changed IP, the new IP is not shown in the peer list:
```
guillaume@potiron:~$ vault operator raft list-peers
Node               Address             State       Voter
----               -------             -----       -----
dev-vault-k8s/0    10.1.182.18:8201    leader      true
dev-vault-k8s/1    10.1.182.57:8201    follower    true
dev-vault-k8s/2    10.1.182.45:8201    follower    true
```

Here are logs from the leader node:

```
2023-11-29T21:59:03.373Z [vault] 2023-11-29T21:59:03.373Z [ERROR] storage.raft: failed to heartbeat to: peer=10.1.182.57:8201 backoff time=2.5s error="dial tcp 10.1.182.57:8201: i/o timeout"
2023-11-29T21:59:04.711Z [vault] 2023-11-29T21:59:04.711Z [WARN]  storage.raft: rejecting vote request since we have a leader: from=10.1.182.19:8201 leader=10.1.182.18:8201 leader-id=dev-vault-k8s/0
2023-11-29T21:59:06.830Z [vault] 2023-11-29T21:59:06.830Z [ERROR] storage.raft: failed to appendEntries to: peer="{Nonvoter dev-vault-k8s/1 10.1.182.57:8201}" error="dial tcp 10.1.182.57:8201: i/o timeout"
2023-11-29T21:59:10.816Z [vault] 2023-11-29T21:59:10.816Z [WARN]  storage.raft: rejecting vote request since we have a leader: from=10.1.182.19:8201 leader=10.1.182.18:8201 leader-id=dev-vault-k8s/0
```

Here are logs from the pod that got deleted:

```
2023-11-29T21:59:33.107Z [vault] 2023-11-29T21:59:33.107Z [WARN]  storage.raft: Election timeout reached, restarting election
2023-11-29T21:59:33.107Z [vault] 2023-11-29T21:59:33.107Z [INFO]  storage.raft: entering candidate state: node="Node at 10.1.182.19:8201 [Candidate]" term=68
2023-11-29T21:59:40.173Z [vault] 2023-11-29T21:59:40.173Z [WARN]  storage.raft: Election timeout reached, restarting election
2023-11-29T21:59:40.173Z [vault] 2023-11-29T21:59:40.173Z [INFO]  storage.raft: entering candidate state: node="Node at 10.1.182.19:8201 [Candidate]" term=69
```

#### Environment

- Charm version: Rev 53 (edge)
- Juju version: 3.1.6
- Cloud Environment: MicroK8s
- Kubernetes version: v1.27.7

--- a/src/charm.py
+++ b/src/charm.py
@@ -336,8 +336,8 @@ class VaultCharm(CharmBase):
         if not self._container.can_connect():
             return
         try:
-            root_token, unseal_keys = self._get_initialization_secret_from_peer_relation()
-            if self._bind_address:
+            root_token, _ = self._get_initialization_secret_from_peer_relation()
+            if self._api_address:
                 vault = Vault(
                     url=self._api_address, ca_cert_path=self._get_ca_cert_location_in_charm()
                 )
@@ -599,6 +599,14 @@ class VaultCharm(CharmBase):
         """
         return f"https://{socket.getfqdn()}:{self.VAULT_PORT}"
 
+    @property
+    def _cluster_address(self) -> str:
+        """Returns the FQDN with the https schema and vault cluster port.
+
+        Example: "https://vault-k8s-1.vault-k8s-endpoints.test.svc.cluster.local:8201"
+        """
+        return f"https://{socket.getfqdn()}:{self.VAULT_CLUSTER_PORT}"
+
     def _push_ca_certificate_to_workload(self, certificate: str) -> None:
         """Push the CA certificate to the workload.
 
@@ -641,7 +649,7 @@ class VaultCharm(CharmBase):
         content = render_vault_config_file(
             default_lease_ttl=self.model.config["default_lease_ttl"],
             max_lease_ttl=self.model.config["max_lease_ttl"],
-            cluster_address=f"https://{self._bind_address}:{self.VAULT_CLUSTER_PORT}",
+            cluster_address=self._cluster_address,
             api_address=self._api_address,
             tcp_address=f"[::]:{self.VAULT_PORT}",
             tls_cert_file=TLS_CERT_FILE_PATH,

--- a/src/charm.py
+++ b/src/charm.py
@@ -336,8 +336,8 @@ class VaultCharm(CharmBase):
         if not self._container.can_connect():
             return
         try:
-            root_token, _ = self._get_initialization_secret_from_peer_relation()
-            if self._api_address:
+            root_token, unseal_keys = self._get_initialization_secret_from_peer_relation()
+            if self._bind_address:
                 vault = Vault(
                     url=self._api_address, ca_cert_path=self._get_ca_cert_location_in_charm()
                 )

--- a/tests/unit/config.hcl
+++ b/tests/unit/config.hcl
@@ -15,7 +15,7 @@ listener "tcp" {
 default_lease_ttl = "168h"
 max_lease_ttl     = "720h"
 disable_mlock     = true
-cluster_addr      = "https://1.2.3.4:8201"
+cluster_addr      = "https://myhostname:8201"
 api_addr          = "https://myhostname:8200"
 telemetry {
   disable_hostname = true

